core/src/verify.rs

```diff
@@ -177,14 +177,61 @@ impl CommitSequenceVerifier {
 
     /// Verifies whether the given reserved state is valid from the current state.
     pub fn verify_reserved_state(&self, _rs: &ReservedState) -> Result<(), Error> {
-        // TODO:
-        // 1. Check that the number of members is at least 4.
-        // 2. Check that the version advances correctly.
-        // 3. Check that `consensus_leader_order` is correct.
-        // 4. Check that `genesis_info` stays the same.
-        // 5. Check that the newly added (if exists) `Member::name` is unique.
-        // 6. Check that `member` monotonicaly increases (refer to `Member::expelled`).
-        // 7. Check that the delegation state doesn't change.
+        // Check that the number of members is at least 4.
+        if _rs.members.len() < 4 {
+            return Err(Error::InvalidArgument(
+                "the number of members is less than 4".to_string(),
+            ));
+        }
+        // Check that `consensus_leader_order` is correct.
+        let mut leader_names = _rs
+            .members
+            .iter()
+            .map(|m| m.name.clone())
+            .collect::<Vec<_>>();
+        leader_names.sort();
+        if leader_names != _rs.consensus_leader_order {
+            return Err(Error::InvalidArgument(
+                "consensus_leader_order is incorrect".to_string(),
+            ));
+        }
+        // Check that `genesis_info` stays the same.
+        if _rs.genesis_info != self.reserved_state.genesis_info {
+            return Err(Error::InvalidArgument("genesis_info changes".to_string()));
+        }
+        // Check that the newly added (if exists) `Member::name` is unique.
+        let existing_names = self
+            .reserved_state
+            .members
+            .iter()
+            .map(|m| &m.name)
+            .collect::<HashSet<_>>();
+        for member in &_rs.members {
+            if !existing_names.contains(&member.name)
+                && _rs.members.iter().filter(|m| m.name == member.name).count() > 1
+            {
+                return Err(Error::InvalidArgument(format!(
+                    "Member name '{}' already exists",
+                    member.name
+                )));
+            }
+        }
+        // Check that `member` monotonically increases (refer to `Member::expelled`).
+        let existing_members = self
+            .reserved_state
+            .members
+            .iter()
+            .map(|m| m.public_key.clone())
+            .collect::<HashSet<_>>();
+        for member in &_rs.members {
+            if existing_members.contains(&member.public_key) && member.expelled {
+                return Err(Error::InvalidArgument(format!(
+                    "Member '{}' cannot be expelled",
+                    member.name
+                )));
+            }
+        }
+        // Check that the delegation state doesn't change.
```

Comment:
What is this comment for?

Reply:
The mistake has happened to missing the validation logic. I added this on the consequent commit.